R/utilities-simulation-results.R

```diff
@@ -0,0 +1,126 @@
+#' @param quantitiesOrPaths Quantity instances (element or vector) (typically retrieved using
+#' \code{getAllQuantitiesMatching}) or quantity path (element or vector of strings) for which the results are to be returned.
+#' When providing the paths, only absolute full paths are supported (i.e., no matching with '*' possible)
+#' @param individualIds \code{numeric} IDs of individiuals for which the results should be extracted.
+#' By default, all individuals from the results are considered. If the individual with the provided ID is not found, the ID is ignored
+#'
+#' @return A list of results, each entry having the name of a quantity path and consisting of the element \code{x} being the time values (in minutes)
+#' and the element \code{y} a \code{NxM} matrix of simulated values, with \code{N} the number of entries (equals the number of output time steps)
+#' and \code{M} the number of individuals. If no individual contains the results for a given path, the value of the element is \code{NULL}
+#'
+#' @seealso \code{\link{getResultsValues}}
+#'
+#' @export
+#'
+#' @examples
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' simResults <- runSimulation(sim)
+#'
+#' paths <- "Organism|Volume"
+#'
+#' resultsValues <- getPopulationResultsValues(simResults, paths, individualIds = 0)
+#' plot(resultsValues[[paths]])
+getPopulationResultsValues <- function(simulationResults, quantitiesOrPaths, individualIds = NULL) {
+  validateIsOfType(simulationResults, "SimulationResults")
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+  validateIsOfType(quantitiesOrPaths, c("Quantity", "character"))
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(quantitiesOrPaths, "Quantity")) {
+    paths <- unlist(lapply(quantitiesOrPaths, function(x) x$consolidatePath))
+  }
+  paths <- unique(paths)
+
+  # If no specific individual ids are passed, iterate through all individuals
+  if (is.null(individualIds)) {
+    individualIds <- simulationResults$allIndividualIds
+  }
+  else {
+    individualIds <- unique(individualIds)
+  }
+
+  # Pre-alloate a list with the number of entries equal to the number of paths
+  output <- vector("list", length(paths))
+  names(output) <- paths
+  # All time values are equal
+  timeVals <- simulationResults$timeValues
+
+  for (path in paths) {
+    # Pre-allocate the output matrix. The number of rows equals the number of simulated time points.
+    # The number of columns is the number of possible individuals.
+    outputValues <- matrix(nrow = length(timeVals), ncol = max(length(individualIds), simulationResults$count), dimnames = NULL)
+
+    for (individualIdx in seq_along(individualIds)) {
+      individualId <- individualIds[individualIdx]
+      vals <- simulationResults$getValuesForIndividual(resultPath = path, individualId = individualId)
+      if (is.null(vals)) {

```

Comment:
That would mean asking for invalid path or results. Should not we throw an exception here? This is clearly an input error no?

Reply:
I am thinking of use cases where the user prepares a set of output paths and re-uses it between simulations/projects. Let's say my default output selection contain 20 output paths. Depending on the model I simulate (e.g. with or without the metabolite), the results might contain only a sub-set of this default output selection. It would be OK from my point of view to just return the results that actually are present and ignore the rest.

Reply:
Ok I am going to leave @KatrinCoboeken and @Yuri05 to decide on that. I think this could cause serious headache in the long run where the user won't understand what's going on because the wrong results are returned or missing.

Reply:
How about a compromise in form of a `stopIfNotFound` argument, with default `TRUE`, as already discussed for getting parameters? In default state, the function would throw an error if an output or an individual id is not found. This behavior could be overridden by setting `stopIfNotFound = FALSE`, then it would behave like it does now.

Reply:
sure. I am going to create another issue for that so that we can merge this guy